core/src/test/java/com/linecorp/armeria/server/AnnotatedValueResolverTest.java

```diff
@@ -0,0 +1,1023 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.server.AnnotatedValueResolver.NoAnnotatedParameterException;
+import com.linecorp.armeria.server.AnnotatedValueResolver.RequestObjectResolver;
+import com.linecorp.armeria.server.AnnotatedValueResolver.ResolverContext;
+import com.linecorp.armeria.server.annotation.Default;
+import com.linecorp.armeria.server.annotation.Header;
+import com.linecorp.armeria.server.annotation.Param;
+import com.linecorp.armeria.server.annotation.RequestObject;
+
+import io.netty.handler.codec.http.QueryStringEncoder;
+import io.netty.util.AsciiString;
+
+public class AnnotatedValueResolverTest {
+    private static final Logger logger = LoggerFactory.getLogger(AnnotatedValueResolverTest.class);
+
+    static final List<RequestObjectResolver> objectResolvers = toRequestObjectResolvers(ImmutableList.of());
+
+    // A string which is the same as the parameter will be returned.
+    static final Set<String> pathParams = ImmutableSet.of("var1");
+    static final Set<String> existingHttpParameters = ImmutableSet.of("param1",
+                                                                      "enum1",
+                                                                      "sensitive");
+
+    // 'headerValues' will be returned.
+    static final Set<AsciiString> existingHttpHeaders = ImmutableSet.of(AsciiString.of("header1"),
+                                                                        AsciiString.of("header2"));
+    static final List<String> headerValues = ImmutableList.of("value1",
+                                                              "value3",
+                                                              "value2");
+
+    static final ResolverContext resolverContext = mock(ResolverContext.class);
+    static final ServiceRequestContext context = mock(ServiceRequestContext.class);
+    static final HttpRequest request = mock(HttpRequest.class);
+    static final AggregatedHttpMessage message = mock(AggregatedHttpMessage.class);
+    static final HttpHeaders headers = mock(HttpHeaders.class);
+    static final HttpParameters parameters = mock(HttpParameters.class);
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        when(resolverContext.httpParameters()).thenReturn(parameters);
+        when(resolverContext.context()).thenReturn(context);
+        when(resolverContext.request()).thenReturn(request);
+        when(resolverContext.message()).thenReturn(message);
+
+        when(context.request()).thenReturn(request);
+        when(context.query())
+                .thenReturn(new QueryStringEncoder(
```

Comment:
Seems like `new QueryStringEncoder()` is not necessary at all or you need to use `QueryStringEncoder.addParam()` for each parameters?